src/java.base/share/classes/java/net/ServerSocket.java

```diff
@@ -78,17 +78,17 @@
  */
 public class ServerSocket implements java.io.Closeable {
     /**
-     * Various states of this socket.
+     * The underlying SocketImpl
      */
-    private boolean created = false;
-    private boolean bound = false;
-    private boolean closed = false;
-    private Object closeLock = new Object();
+    private final SocketImpl impl;
 
     /**
-     * The implementation of this Socket.
+     * Various states of this socket, need stateLock to change.
      */
-    private SocketImpl impl;
+    private volatile boolean created;   // impl.create(boolean) called
+    private volatile boolean bound;
+    private volatile boolean closed;
```

Comment:
I don't see why these two need to be `volatile`, but we can keep them for extra safety.

Reply:
isBound and isClose poll the state.

Reply:
Ah yes, I see. Doing `while (!socket.isClosed()) ...` would be problematic otherwise.